### ISSUE TITLE
disabling auto-completion in obvious places

### ISSUE THREAD

**Issue by leodemoura** [leanprover/lean#150]
@soonhokong I just noticed an annoying behavior. I was writing a new theorem

```
        theorem decida
```

And auto-completion started showing a bunch of alternatives. They were clearly irrelevant, since I'm writing the name of a new theorem. Can we fix that? The idea would be to disable auto-completion in places where it is clearly irrelevant (e.g., after a theorem/definition/parameter/... keyword).
Another place it is is irrelevant is when we are writing the name of parameters. However, it will be much harder to detect that in Emacs.

@avigad and @fpvandoorn : what do you think?

**Comment by soonhokong** [comment]
@leodemoura, this is certainly possible.

I'm writing a function `company-lean--need-autocomplete` which uses a regular-expression to determine whether lean-mode needs autocomplete or not. The following is a draft that I have. Would you give me more entries?

``` elisp
(defun company-lean--need-autocomplete ()
  (not (looking-back
        (rx (or "theorem"
                "definition"
                "lemma"
                "axiom"
                "parameter"
                "abbreviation")
            (+ white)
            (* (not white))))))
```

**Comment by leodemoura** [comment]
Here are additional entries:  "variable", "hypothesis", "conjecture" "corollary"

**Comment by soonhokong** [comment]
@leodemoura, thanks. I'll add them and commit.

> Another place it is is irrelevant is when we are writing the name of parameters.

Do you think it's difficult to filter this one using regular expression?

**Comment by leodemoura** [comment]
> Do you think it's difficult to filter this one using regular expression?

I think we could try easy cases, but it will be hard to get a precise solution using only regular expressions. For example, in principle, we want to define auto-completion in the binder names.
Binders occur in many places: Pi, lambda, forall, after declarations, and in user defined notation.
Let's consider the lambda expression

```
        fun (id_1 : type_1) ... (id_n : type_n), ...
```

We want to disable auto-completion for all `id_i`s.

**Comment by soonhokong** [comment]
I see.

For the complicated cases, I think we need to extend `FINDP` command to take a column-number in addition to line-number and handle them from lean-server side..

**Comment by leodemoura** [comment]
I agree.

**Comment by soonhokong** [comment]
Closed by c365f6b9ab67fc6ca31cbe6911e49ad4ff980c79.
